File: src/keboola_mcp_server/tools/search/usage.py
from collections import defaultdict
from typing import Mapping, Optional, Sequence

from pydantic import BaseModel, Field

from keboola_mcp_server.clients.base import JsonStruct
from keboola_mcp_server.clients.client import (
    KeboolaClient,
    get_metadata_property,
)
from keboola_mcp_server.config import MetadataField
from keboola_mcp_server.tools.search.tools import SearchItemType, SearchSpec, _fetch_configurations


class ComponentUsageReference(BaseModel):
    component_id: str = Field(description='The ID of the component.')
    configuration_id: str = Field(description='The ID of the configuration.')
    configuration_row_id: str | None = Field(default=None, description='The ID of the configuration row.')
    configuration_name: str | None = Field(default=None, description='The name of the configuration.')
    used_in: str | None = Field(default=None, description='The scope in which the target item was used.')
    timestamp: str | None = Field(default=None, description='The timestamp of the usage.')


class UsageById(BaseModel):
    target_id: str
    usage_references: list[ComponentUsageReference]


async def find_id_usage(
    client: KeboolaClient,
    target_ids: Sequence[str],
    item_types: Optional[Sequence[SearchItemType]] = None,
    scopes: list[str] = list,
) -> list[UsageById]:
    """
    Finds component configurations (including rows) that reference any of the target IDs in the specified configuration
    scopes.

    :param client: The Keboola client to use.
    :param target_ids: The IDs to search for.
    :param item_types: The item types to search for, if not provided, all item types are searched.
    :param scopes: Dot-separated keys to search in the configuration.
    :return: A list of UsageById objects.
    """

    spec = SearchSpec(
        patterns=target_ids,
        item_types=item_types or tuple(),
        search_scopes=scopes,
        pattern_mode='literal',
        search_type='config-based',
        return_all_matched_patterns=True,
        stop_searching_after_first_value_match=False,
    )

    search_hits = await _fetch_configurations(client, spec)

    # group usage references by pattern = target_id
    output: dict[str, list[ComponentUsageReference]] = defaultdict(list)
    for search_hit in search_hits:
        for match in search_hit._matches:
            for pattern in match.patterns:
                output[pattern].append(
                    # TODO: Consider whether adding configuration description is useful, it could overload context.
                    ComponentUsageReference(
                        component_id=search_hit.component_id,
                        configuration_id=search_hit.configuration_id,
                        configuration_row_id=search_hit.configuration_row_id,
                        configuration_name=search_hit.name,
                        used_in=match.scope,
                        timestamp=search_hit.updated,
                    )
                )
    return [
        UsageById(target_id=target_id, usage_references=usage_references)
        for target_id, usage_references in output.items()
    ]


def get_created_by(
    metadata: Sequence[Mapping[str, JsonStruct]] | Mapping[str, JsonStruct],
) -> ComponentUsageReference | None:
    """
    Gets the created by reference from the metadata.
    :param metadata: The metadata to search in.
    :return: The created by reference.
    """
    metadata_items = _coerce_metadata_list(metadata)
    component_id = get_metadata_property(metadata_items, MetadataField.CREATED_BY_COMPONENT_ID)
    configuration_id = get_metadata_property(metadata_items, MetadataField.CREATED_BY_CONFIGURATION_ID)
    configuration_row_id = get_metadata_property(metadata_items, MetadataField.CREATED_BY_CONFIGURATION_ROW_ID)
    timestamp = _get_latest_metadata_timestamp(
        metadata_items,
        [
            MetadataField.CREATED_BY_COMPONENT_ID,
            MetadataField.CREATED_BY_CONFIGURATION_ID,
            MetadataField.CREATED_BY_CONFIGURATION_ROW_ID,
        ],
    )
    if component_id is None or configuration_id is None or timestamp is None:
        return None
    return ComponentUsageReference(
        component_id=str(component_id),
        configuration_id=str(configuration_id),
        configuration_row_id=str(configuration_row_id) if configuration_row_id else None,
        configuration_scope=None,
        timestamp=timestamp,
    )


def get_last_updated_by(
    metadata: Sequence[Mapping[str, JsonStruct]] | Mapping[str, JsonStruct],
) -> ComponentUsageReference | None:
    """
    Gets the last updated by reference from the metadata.
    :param metadata: The metadata to search in.
    :return: The last updated by reference.
    """
    metadata_items = _coerce_metadata_list(metadata)
    component_id = get_metadata_property(metadata_items, MetadataField.UPDATED_BY_COMPONENT_ID)
    configuration_id = get_metadata_property(metadata_items, MetadataField.UPDATED_BY_CONFIGURATION_ID)
    configuration_row_id = get_metadata_property(metadata_items, MetadataField.UPDATED_BY_CONFIGURATION_ROW_ID)
    timestamp = _get_latest_metadata_timestamp(
        metadata_items, [MetadataField.UPDATED_BY_COMPONENT_ID, MetadataField.UPDATED_BY_CONFIGURATION_ID]
    )
    if component_id is None or configuration_id is None or timestamp is None:
        return None
    return ComponentUsageReference(
        component_id=str(component_id),
        configuration_id=str(configuration_id),
        configuration_row_id=str(configuration_row_id) if configuration_row_id else None,
        configuration_scope=None,
        timestamp=timestamp,
    )


def _coerce_metadata_list(
    metadata: Sequence[Mapping[str, JsonStruct]] | Mapping[str, JsonStruct],
) -> list[Mapping[str, JsonStruct]]:
    if isinstance(metadata, Mapping):
        metadata_value = metadata.get('metadata')
        if isinstance(metadata_value, list):
            return [item for item in metadata_value if isinstance(item, Mapping)]
        if 'key' in metadata and 'value' in metadata:
            return [metadata]
        return []
    return [item for item in metadata if isinstance(item, Mapping)]


def _get_latest_metadata_timestamp(metadata: list[Mapping[str, JsonStruct]], keys: Sequence[str]) -> str | None:
    latest = None
    for key in keys:
        for item in metadata:
            if item.get('key') != key:
                continue
            timestamp = item.get('timestamp')
            if timestamp is None:
                continue
            if latest is None or str(timestamp) > str(latest):
                latest = str(timestamp)
    return latest

File: src/tools/storage/usage.ts
// Ported from tools/storage/usage.py (get_created_by / get_last_updated_by).
//
// find_id_usage depends on the search subsystem (tools/search.py), which is not yet
// ported to TypeScript. include_usage therefore returns empty usage; see tools.ts.

import { getMetadataProperty, parseIsoTimestamp, type RawObj } from './model';

const CREATED_BY_COMPONENT_ID = 'KBC.createdBy.component.id';
const CREATED_BY_CONFIGURATION_ID = 'KBC.createdBy.configuration.id';
const CREATED_BY_CONFIGURATION_ROW_ID = 'KBC.createdBy.configurationRow.id';
const UPDATED_BY_COMPONENT_ID = 'KBC.lastUpdatedBy.component.id';
const UPDATED_BY_CONFIGURATION_ID = 'KBC.lastUpdatedBy.configuration.id';
const UPDATED_BY_CONFIGURATION_ROW_ID = 'KBC.lastUpdatedBy.configurationRow.id';

export type ComponentUsageReference = {
  component_id: string;
  configuration_id: string;
  configuration_row_id: string | null;
  configuration_name: string | null;
  used_in: string | null;
  timestamp: string | null;
};

const latestMetadataTimestamp = (metadata: RawObj[], keys: string[]): string | null => {
  let latest: number | null = null;
  let latestRaw: string | null = null;
  for (const item of metadata) {
    if (!keys.includes(item.key as string)) continue;
    const rawTs = item.timestamp;
    if (typeof rawTs !== 'string') continue;
    let parsed: number;
    try {
      parsed = parseIsoTimestamp(rawTs);
    } catch {
      continue;
    }
    if (latest === null || parsed > latest) {
      latest = parsed;
      latestRaw = rawTs;
    }
  }
  return latestRaw;
};

const lineageReference = (
  metadata: unknown,
  componentKey: string,
  configKey: string,
  rowKey: string,
): ComponentUsageReference | null => {
  if (!Array.isArray(metadata)) return null;
  const items = metadata as RawObj[];
  const componentId = getMetadataProperty(items, componentKey);
  const configurationId = getMetadataProperty(items, configKey);
  const rowId = getMetadataProperty(items, rowKey);
  if (componentId === null || configurationId === null) return null;
  return {
    component_id: String(componentId),
    configuration_id: String(configurationId),
    configuration_row_id: rowId ? String(rowId) : null,
    configuration_name: null,
    used_in: null,
    timestamp: latestMetadataTimestamp(items, [componentKey, configKey, rowKey]),
  };
};

export const getCreatedBy = (metadata: unknown): ComponentUsageReference | null =>
  lineageReference(
    metadata,
    CREATED_BY_COMPONENT_ID,
    CREATED_BY_CONFIGURATION_ID,
    CREATED_BY_CONFIGURATION_ROW_ID,
  );

export const getLastUpdatedBy = (metadata: unknown): ComponentUsageReference | null =>
  lineageReference(
    metadata,
    UPDATED_BY_COMPONENT_ID,
    UPDATED_BY_CONFIGURATION_ID,
    UPDATED_BY_CONFIGURATION_ROW_ID,
  );
